R/register_extendr.R

```diff
@@ -40,27 +45,38 @@ register_extendr <- function(path = ".", quiet = FALSE, force_wrappers = FALSE,
 
   outfile <- rprojroot::find_package_root_file("R", "extendr-wrappers.R", path = path)
 
-  # If force_wrappers is TRUE, use a minimal wrapper file even when
-  # make_wrappers() fails; since the wrapper generation depends on the compiled
-  # Rust code, the package needs to be installed before attempting this, but
-  # it's not always the case (e.g. the package might be corrupted, or not
-  # installed yet).
-  if (isTRUE(force_wrappers)) {
-    error_handle <- function(e) {
-      cli::cli_alert_danger("Failed to generate wrapper functions: {e$message}.")
-      cli::cli_alert_warning("Falling back to a minimal wrapper file instead.")
-
-      make_example_wrappers(pkg_name, outfile, path = path)
-    }
-  } else {
-    error_handle <- function(e) {
-      ui_throw(
-        "Failed to generate wrapper functions.",
-        c(
-          ui_x(e[["message"]])
-        )
-      )
-    }
+  path <- rprojroot::find_package_root_file(path = path)
+
+  # If compile is NA, compile if the DLL is newer than the source files
+  if (isTRUE(is.na(compile))) {
+    compile <- needs_compilation(path, quiet) || pkgbuild::needs_compile(path)
+  }
+
+  if (isTRUE(compile)) {
+    # This relies on [`pkgbuild::needs_compile()`], which
+    # does not know about Rust files modifications.
+    # `force = TRUE` enforces compilation.
+    pkgbuild::compile_dll(
+      path = path,
+      force = TRUE,
+      quiet = quiet
+    )
+  }
+
+  dll_file <- get_library_path(path)
+
+  # If the wrapper file is newer than the DLL file, assume it's generated by
+  # the latest DLL, which should mean it doesn't need regenerated. This isn't
+  # always the case (e.g. when the user accidentally edited the wrapper file
+  # by hand) so the user might need to run with `force = TRUE`.
+  if (!isTRUE(force) &&
+      isTRUE(file.exists(dll_file)) &&
+      isTRUE(file.exists(outfile)) &&
+      isTRUE(file.info(dll_file)$mtime < file.info(outfile)$mtime)) {
+    cli::cli_alert_info("{.file {outfile}} is up-to-date. Skip generating wrapper functions.")
+    cli::cli_alert_info("If you want to force the generation, run `register_extendr(force = TRUE)`.")
```

Comment:
1. I suggest using `cli::cli_alert_warning` to draw the user's attention, as this message is important.
2. I suggest `"..., run {.code rextendr::register_extendr(force = TRUE)},"` for consistency, see [{cli} inline markup](https://cli.r-lib.org/reference/inline-markup.html). Note that `{.fun ...}` will not work, as it appends an extra parentheses to the function name.

Reply:
On second thought, do we really need to write this every time `rextendr::document()` is called without any changes to the Rust code? I'd say this is a standard situation.

Reply:
Thanks, I'm not yet familiar with cli's expressions. Let's use it.

> On second thought, do we really need to write this every time `rextendr::document()` is called without any changes to the Rust code?

Do we want to warn once a session? Or, do we want to stop warning at all?

Reply:
It feels like we do not need it all. @clauswilke, what do you think?

Reply:
I always get annoyed when code tells me stuff I know already. Even if only once a session. I say remove it. People can read the documentation or ask a question on stackoverflow if they truly don't know how to force the generation of the wrappers.